File: leet/leet_611.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        int len = nums.size();
        if(len < 3) return 0;
        /*int cnt = 0;
        // The follows not work: Time Limit Exceeded
        for(int i = 0; i < len - 2; ++i)
        {
            for(int j = i + 1; j < len - 1; ++j)
            {
                for(int k = j + 1; k < len; ++k)
                {
                    if(nums[i] + nums[j] > nums[k] && nums[j] + nums[k] > nums[i] && nums[k] + nums[i] > nums[j]) 
                        ++cnt;
                }
            }
        }
        return cnt;*/
        // sort in ascending order
        sort(nums.begin(), nums.end());
        //if(len == 3) return (nums[0] + nums[1] > nums[2])? 1: 0;
        int cnt = 0, i, j;
        // i points to largest element to be compared
        for (int k = len - 1; k >= 2; --k)
        {
            //if(nums[i] == 0) continue;
            i = 0, j = k - 1;
            // inspired by java(O(n^2) time in O(1) space!!!
            while(i < j)
            {
                if(nums[i] + nums[j] > nums[k])
                {
                    cnt += j - i;
                    --j;
                }
                else
                {
                    ++i;
                }
            }
        }
        return cnt;
    }
};

int main()
{
    Solution solu;
    int a[] = {2,2,2, 3,4};
    vector<int> arr(a, a + sizeof(a)/sizeof(a[0]));
    cout << solu.triangleNumber(arr) << endl;
    return 0;
}

File: leet/leet_611.py
# count the number of triplets from the array that can make triangles 
class Solution(object):
    def triangleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        leng = len(nums)
        if leng < 3:
            return 0
        nums.sort() # in ascending order
        cnt = 0
        for k in range(leng-1, 1, -1):
            i = 0
            j = k - 1
            # guarantee that the middle edge and short edge are different and their combine is unique
            while i < j:
                # if as small as i is ok, then all i till j-1 is ok (total j-1-i+1)
                if nums[i] + nums[j] > nums[k]:
                    cnt += j - i
                    j -= 1 # decreasing the middle edge anchor
                else:
                    i += 1
        return cnt

if __name__ == '__main__':
    solu = Solution()
    print solu.triangleNumber([2,2,2,3,4])
